### PR TITLE
snap: make `snap download` download via snapd if available

### PR DESCRIPTION
**Pull request by mvo5**
The `snap download` command will now try to talk to snapd
directly. There is also an additional "--direct" switch now to
force going to the store directly and it will automatically fallback
if no snapd is available.

Based on #7622

### CODE REVIEW

File: daemon/api_download.go [hunk 1]
```diff
@@ -68,7 +68,7 @@ func streamOneSnap(c *Command, action snapDownloadAction, user *auth.UserState)
 	}}
 	snaps, err := getStore(c).SnapAction(context.TODO(), nil, actions, user, nil)
 	if err != nil {
-		return errToResponse(err, []string{action.SnapName}, InternalError, "cannot download snap: %v")
+		return errToResponse(err, []string{action.SnapName}, InternalError, "%v")
```
Comment (pedronis) [inline daemon/api_download.go]: why this change?

Comment (mvo5) [reply daemon/api_download.go]: We currently get the following error when a download fails:
```
error: cannot download snap: cannot download snap "fdfasd": snap not found
```
(note the duplication of "cannot download snap").

The store errors.go code already adds the "cannot download snap" prefix in SnapActionError.Error() (`"cannot %s snap %q: %v", op, name, e`). But given that this is cosmetic I can move it out of this PR and put it into a separate one given that its a bit distracting here.

File: cmd/snap/cmd_download.go [hunk 2]
```diff
@@ -170,3 +159,141 @@ func (x *cmdDownload) Execute(args []string) error {
...
+	if err := fetcher.Fetch(ref); err != nil {
+		return err
+	}
+	// XXX: cross check?
+
+	return nil
+}
+
+func isErrorKindLoginRequired(e error) bool {
+	if err, ok := e.(*client.Error); ok {
```
Comment (pedronis) [inline cmd/snap/cmd_download.go]: can't we use As for part of this as well?

File: cmd/snap/cmd_download.go [hunk 3]
```diff
@@ -170,3 +159,141 @@ func (x *cmdDownload) Execute(args []string) error {
...
+	}
+	snapName := string(x.Positional.Snap)
+
+	if x.Direct {
+		return x.downloadDirect(snapName, revision)
+	}
+
+	err := x.downloadViaSnapd(snapName, revision)
+	if isConnectionError(err) || isErrorKindLoginRequired(err) {
+		fmt.Fprintf(Stdout, "Cannot connect to the daemon, trying direct download\n")
```
Comment (pedronis) [inline cmd/snap/cmd_download.go]: do we really want to print this? should it go to Stderr?  even if we do, it should be probably be prefix, as some thing in `snap info`,  also missing i18n ?

File: cmd/snap/cmd_download.go [hunk 4]
```diff
@@ -170,3 +159,142 @@ func (x *cmdDownload) Execute(args []string) error {
...
+	if err := os.MkdirAll(filepath.Dir(downloadPath), 0755); err != nil {
+		return err
+	}
+	f, err := os.Create(downloadPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if _, err := io.Copy(f, stream); err != nil {
```
Comment (chipaca) [inline cmd/snap/cmd_download.go]: this gives you no progress bar (you'll have to tweak `client.Download` to get the download size here! (it's `rsp.ContentLength`, right next to `rsp.Body` one layer up; that omission and the loss of 'resume' I should have spotted earlier, sorry), and should also be doing the sha3 checksumming.
Look at store's Download to see how to set up a multiwriter for this
Look at wait's signal handling to see how to set things up for the progress bar to dtrt in the face of signals

File: cmd/snap/cmd_download.go [hunk 5]
```diff
@@ -170,3 +159,142 @@ func (x *cmdDownload) Execute(args []string) error {
...
+		headers, err := asserts.HeadersFromPrimaryKey(ref.Type, ref.PrimaryKey)
+		if err != nil {
+			return nil, err
+		}
+		asserts, err := x.client.Known(ref.Type.Name, headers, &client.KnownOptions{Remote: true})
+		if err != nil {
+			return nil, err
+		}
+		if len(asserts) != 1 {
+			return nil, fmt.Errorf("unexpected amount of assertions: %q", asserts)
```
Comment (chipaca) [inline cmd/snap/cmd_download.go]:
```suggestion
			return nil, fmt.Errorf("unexpected number of assertions: expected 1, got %d", len(asserts))
```

File: cmd/snap/cmd_download.go (hunk 4 above)
Comment (chipaca) [reply cmd/snap/cmd_download.go]: Also, the store told you what the sha3 *should* be, and we discarded that information. Maybe it should be set in a header in the response, in `store.DownloadStream`, so that the client can pick it up and check it before hitting the assertions db.

Comment (chipaca) [reply cmd/snap/cmd_download.go]: We *do* have the size! `response.fileStream` seems to DTRT with it.

Comment (pedronis) [reply cmd/snap/cmd_download.go]: @chipaca @mvo idea, maybe expose the store.go downloadImpl helper is some form ?

File: cmd/snap/cmd_download.go [hunk 6]
```diff
@@ -170,3 +163,160 @@ func (x *cmdDownload) Execute(args []string) error {
...
+		stream.Close()
+	}()
+	defer signal.Reset(syscall.SIGINT)
+
+	h := crypto.SHA3_384.New()
+	mw := io.MultiWriter(f, h, pbar)
+	if _, err := io.Copy(mw, stream); err != nil {
+		return err
+	}
+	pbar.Finished()
```
Comment (chipaca) [inline cmd/snap/cmd_download.go]: shouldn't you also `stream.Cose()` here?

File: cmd/snap/cmd_download.go [hunk 7]
```diff
@@ -170,3 +163,160 @@ func (x *cmdDownload) Execute(args []string) error {
...
+
+	fmt.Fprintf(Stdout, i18n.G("Fetching assertions for %q\n"), snapName)
+
+	assertFname := strings.TrimSuffix(downloadPath, filepath.Ext(downloadPath)) + ".assert"
+	sha3_384, err := asserts.EncodeDigest(crypto.SHA3_384, h.Sum(nil))
```
Comment (chipaca) [inline cmd/snap/cmd_download.go]: can you add a `// TODO` to compare this hash with the one the store provided us?

File: cmd/snap/cmd_download.go (hunk 4 above)
Comment (chipaca) [reply cmd/snap/cmd_download.go]: possible there's a way to split out the bits we want to reuse from `downloadImpl` into their own helper that can then be exposed, but it's not immediately obvious to me today

File: cmd/snap/cmd_download.go (hunk 7 above)
Comment (mvo5) [reply cmd/snap/cmd_download.go]: The code now checks the hash.

**Comment by pedronis** [comment]
@mvo5 @chipaca 

On a system were you are not logged in or don't use sudo I get a polkit prompt dialog, I suspect that might not be what we want here (is very script unfriendly and scripts would have worked before).

Maybe we need to tweak the auth rules for /download alone.

**Comment by pedronis** [comment]
```
var snapDownloadCmd = &Command{
	Path:     "/v2/download",
	PolkitOK: "io.snapcraft.snapd.manage",
	POST:     postSnapDownload,
}
```

should we just drop the PolkitOK line there?

File: cmd/snap/cmd_download.go [hunk 8]
```diff
@@ -97,6 +127,189 @@ func fetchSnapAssertions(tsto *image.ToolingStore, snapPath string, snapInfo *sn
...
+	}
+	printInstallHint(downloadPath, assertFname)
+
+	return nil
+}
+
+func isErrorKindLoginRequired(err error) bool {
+	var clientErr *client.Error
+	if xerrors.As(err, &clientErr) {
+		return clientErr.Kind == client.ErrorKindLoginRequired
```
Comment (pedronis) [inline cmd/snap/cmd_download.go]: this bit doesn't seem to be tested

File: cmd/snap/cmd_download.go [hunk 9]
```diff
@@ -64,14 +76,32 @@ func init() {
...
+	if p, err := filepath.Rel(wd, snapPath); err == nil {
+		snapPath = p
+	}
+	// add a hint what to do with the downloaded snap (LP:1676707)
+	fmt.Fprintf(Stdout, i18n.G(`Install the snap with:
+   snap ack %s
+   snap install %s
+`), assertPath, snapPath)
+}
+
```
Comment (pedronis) [inline cmd/snap/cmd_download.go]: We might want a TODO in this file about trying to share more code with and look at reorganizing image helpers. Notice that in this factoring of things ubuntu-image doesn't work like snap download.

File: cmd/snap/cmd_download.go [hunk 10]
```diff
@@ -97,6 +127,189 @@ func fetchSnapAssertions(tsto *image.ToolingStore, snapPath string, snapInfo *sn
...
+	pbar := progress.MakeProgressBar()
+	pbar.Start(filepath.Base(downloadPath), float64(dlInfo.Size))
+	defer pbar.Finished()
+
+	// Intercept sigint
+	c := make(chan os.Signal, 3)
+	signal.Notify(c, syscall.SIGINT)
+	go func() {
+		<-c
```
Comment (bboozzoo) [inline cmd/snap/cmd_download.go]: Isn't this leaking a goroutine here? Looked briefly at `signal.Reset()` but I didn't see any code that would close the channel passed to `signal.Notify()`.

**Comment by mvo5** [comment]
Marking for 2.43 - either it goes in for that or if it does not we need to undo the "snap known --remote" default, ie. "snap known --remote" and "snap download" should be consistent in either going via snapd by default or going direct by default.

File: cmd/snap/cmd_download.go (hunk 10 above)
Comment (chipaca) [reply cmd/snap/cmd_download.go]: the goroutine will continue to run until the program gets a SIGINT or it exits, which is what we want, so I think it's ok in this case.

Comment (bboozzoo) [reply cmd/snap/cmd_download.go]: Also the way of using Close() to break io.Copy() can be a bit awkward, and leads to this:
```
$ ./test-snap download core
Fetching snap "core"
core_7917.snap      0%  206kB/s 7m33s
### ^C
error: read unix @->/run/snapd.socket: use of closed network connection
```

Tweaked this a little:
```
diff --git a/cmd/snap/cmd_download.go b/cmd/snap/cmd_download.go
index c40ae7bc83..93fe7c1837 100644
--- a/cmd/snap/cmd_download.go
+++ b/cmd/snap/cmd_download.go
@@ -184,18 +184,28 @@ func downloadSnapFromStreamWithProgress(downloadPath string, stream io.ReadClose
        defer pbar.Finished()
 
        // Intercept sigint
-       c := make(chan os.Signal, 3)
+       c := make(chan os.Signal, 1)
        signal.Notify(c, syscall.SIGINT)
+       done := make(chan struct{}, 1)
+       canceled := false
        go func() {
-               <-c
-               pbar.Finished()
-               stream.Close()
+               select {
+               case <-c:
+                       canceled = true
+                       pbar.Finished()
+                       stream.Close()
+               case <-done:
+               }
        }()
+       defer close(done)
        defer signal.Reset(syscall.SIGINT)
 
        h := crypto.SHA3_384.New()
        mw := io.MultiWriter(f, h, pbar)
        if _, err := io.Copy(mw, stream); err != nil {
+               if canceled {
+                       return fmt.Errorf("download canceled")
+               }
                return err
        }
 
```

Now we get a nicer message:
```
$ snap download core
Fetching snap "core"
core_7917.snap     1%  602kB/s 2m34s
### ^C
error: download canceled
```

**Comment by zyga** [comment]
@mvo5 this needs de-conflicting.

**Comment by sd-hd** [comment]
A merged branch is here:
https://github.com/sd-hd/snapd/tree/snap-donwload-via-snapd-2

I also changed the default to direct and "--indirect" is now an option.

Not sure, if all test-cases are handled. (For me the source has still many unkown places ;-) )

**Comment by mvo5** [comment]
Closing this in favor of #7977 which fully includes this one
